physx/src/controller.rs

```diff
@@ -0,0 +1,115 @@
+#![warn(clippy::all)]
+#![warn(rust_2018_idioms)]
+
+use crate::px_type::*;
+use crate::rigid_dynamic::RigidDynamic;
+
+use glam::Vec3;
+use physx_macros::*;
+
+use physx_sys::{
+    phys_PxCreateControllerManager, PxCapsuleControllerDesc, PxCapsuleControllerDesc_delete,
+    PxCapsuleControllerDesc_isValid, PxCapsuleControllerDesc_new_alloc, PxController,
+    PxControllerDesc, PxControllerManager, PxControllerManager_createController_mut,
+    PxControllerManager_release_mut, PxController_getActor, PxController_getPosition,
+    PxController_release_mut, PxController_setPosition_mut, PxExtendedVec3, PxMaterial, PxScene,
+};
+use thiserror::Error;
+
+#[physx_type]
+impl ControllerDesc {}
+
+#[physx_type]
+impl ControllerManager {
+    pub fn new(scene: *mut PxScene) -> Self {
+        let ptr = unsafe { phys_PxCreateControllerManager(scene, false) };
+        ControllerManager { ptr }
+    }
+
+    pub fn create_controller(&self, desc: &mut ControllerDesc) -> Controller {
+        let controller =
+            unsafe { PxControllerManager_createController_mut(self.ptr, desc.get_raw()) };
+        Controller::new(controller)
+    }
+
+    pub unsafe fn release(self: Self) {
+        PxControllerManager_release_mut(self.ptr);
+    }
+}
+
+#[physx_type(inherit = "ControllerDesc")]
+impl CapsuleControllerDesc {
+    pub fn new(
+        height: f32,
+        radius: f32,
+        step_offset: f32,
+        material: *mut PxMaterial,
+    ) -> Result<Self, ControllerError> {
+        unsafe {
+            let c = PxCapsuleControllerDesc_new_alloc();
+            (*c).height = height;
+            (*c).radius = radius;
+            (*c).stepOffset = step_offset;
+            (*c).material = material;
+
+            if PxCapsuleControllerDesc_isValid(c) {
+                Ok(CapsuleControllerDesc { ptr: c })
+            } else {
+                PxCapsuleControllerDesc_delete(c);
+                Err(ControllerError::InvalidDescription())
+            }
+        }
+    }
+
+    pub fn release(self: Self) {
+        unsafe { PxCapsuleControllerDesc_delete(self.ptr) };
+    }
+}
+
+#[physx_type]
+impl Controller {
+    pub fn new(controller: *mut PxController) -> Self {
+        Self { ptr: controller }
+    }
+
+    pub fn set_position(&mut self, position: Vec3) {
+        unsafe {
+            PxController_setPosition_mut(self.get_raw_mut(), &to_extended(&position));
+        }
+    }
+
+    pub fn get_position(&self) -> Vec3 {
+        unsafe { from_extended(*PxController_getPosition(self.get_raw())) }
+    }
+
+    pub fn get_actor(&self) -> RigidDynamic {
+        unsafe { RigidDynamic::new(PxController_getActor(self.ptr)) }
```

Comment:
use `from_ptr` here instead. They have slightly different semantics that we should probably make more apparent. `new` will allocate a Box of UserData, while `from_ptr` does not. :)